### DISCUSSION TITLE
Keycloak Theming Capabilities Using Modern Front End Frameworks

### DISCUSSION THREAD

**Discussion by fadi-william** [keycloak/keycloak#9093] [Ideas]
Hello!

Thank you for your efforts in making Keycloak better.

I actually like Keycloak. However, there was always that theming/design challenge to match the application's design system integrating with Keycloak.

I wished if there was an option to build the login pages using a modern front-end framework. For example React, Angular, Vue, Svelte, etc.

I believe that this would be possible if the architecture of this part has been changed from an MVC architecture to a Web API architecture. Maybe, for security concerns, you could allow us to configure CORS rules for this API access to match the desired front-end hosting requirements.

If this did change. The next thing would be to create different responsive themes for the most commonly used design systems by the different clients.

1. Material Design for Android.
2. Cupertino for iOS.
3. Modern Patternfly Implementation.

I would prefer this to be implemented using a modular swappable components architecture. For example, define a button component with a specific interface (props in react). The implementor will need to just implement a button with the same interface to customize the look of the button. 

I will be more than happy to contribute to make this happen if you just provided me with directions as the Keycloak code is a black box for me currently.

Many thanks for your considerations.

Best Regards.

Fadi

**Comment by ssilvert** [comment]
Thanks for the offer of contributions.  We do plan to build a new version
of the Login theme, probably based on React.  However, I don't have a
timeline for this project yet.

As this is an extremely mission critical part of Keycloak, we probably need
to wait until we can do a comprehensive analysis of requirements.  At that
point we would welcome help and feedback from the community.

Stan

**Reply by fadi-william** [reply]
More than welcome @ssilvert!

Thanks for letting me know that.

Sure. I understand 👍

For the time being. I plan to use https://github.com/InseeFrLab/keycloakify. However, I wished if there was a native solution supported by KeyCloak. 

React is good and will be enough for me. However, in my opinion, allowing multiple frameworks will help developers use their own design systems without needing to switch to another framework or re-implement some of their components.

Have a great day.

Best Regards.

Fadi

**Reply by garronej** [reply]
Hi @ssilvert,  
I am the original author of https://keycloakify.dev.  
Keycloakify is now quite mature it would be a shame this community effort was discarded.  

If you are interested, instead of rebuilding the theming system from scratch, there are some very little changes
that Keycloak can do that would make Keycloakify's work much easier.  
This approach would enable the community to easily come up with equivalent of Keycloakify for Vue, Angular, Svelte ect.  
You would empower the community to build the tooling they need while you can focus on Keycloak's core features.  

I am very interested in collaborating with you on that,

If you are interested please do not hesitate to reach out.  

Best regards,

**Comment by ssilvert** [comment]
@garronej I'm definitely interested.  We will start looking at these issues in Q3 of this year.

**Reply by garronej** [reply]
Hi @ssilvert,  

I just wanted to let you know that I will soon release v6 of Keycloakify which features:  
- A real type safe i18n customization API.  
- Smaller bundle size, faster first print: Heavy use of React Suspence and dynamic imports in particular for only downloading the required locals.  
- A better documentation.  

You can check the v6 documentation website here:  https://docs.keycloakify.dev/v/v6/how-to-use  

I would be interested in knowing what are your future plans about customizability.   
As I said, I'm very willing to keep maintaining and improving Keyclaokify for the foreseeable future. I think we can achieve great things with only minimal involvement on your side.  

Best regards ,

**Comment by ssilvert** [comment]
Thanks @garronej.  We are starting to think about this topic a lot more.  We will definitely take a close look at Keycloakify.

**Comment by PinPinIre** [comment]
@ssilvert It looks like the [account-ui and admin UI](https://github.com/keycloak/keycloak-ui/tree/main/apps) have been migrated to React using Patternfly? Will the plan over the coming year, be to continue with this pattern and port the login pages to React and Patternfly?

**Reply by garronej** [reply]
Just to say that we have a plan to support account and admin theming in [Keycloakify](https://www.keycloakify.dev/).  
Ongoing discussion: https://github.com/InseeFrLab/keycloakify/issues/158

**Reply by ssilvert** [reply]
> @ssilvert It looks like the [account-ui and admin UI](https://github.com/keycloak/keycloak-ui/tree/main/apps) have been migrated to React using Patternfly? Will the plan over the coming year, be to continue with this pattern and port the login pages to React and Patternfly?

Probably.

**Reply by PinPinIre** [reply]
@ssilvert I was wondering if there are plans to migrate login soon or is it still something that probably might happen? Up to now I have been making minimal changes to existing login themes to avoid issues in future when we would have to migrate the login pages to React.

**Reply by garronej** [reply]
I sincerely hope there are no plans to deprecate the current login theme system. Thanks to Keycloakify, the development experience for creating React-based login themes has been significantly improved. 
@ssilvert, would it be possible to consider scheduling a meeting?  
I believe a discussion could be highly beneficial for ensuring that the Keycloakify community and the core Keycloak team are aligned in their efforts. Thank you for your time and consideration.

**Reply by jonkoops** [reply]
Considering the login theme is one of the most often customized parts of Keycloak, we will take it under very careful consideration before making any changes that could be disruptive for our users. We should have done the same for the Admin Console in regards to customization, but were under tremendous pressure to deliver the new version, so things had to budge.

If anything, we want to learn from our mistakes there. So we'd like to invite the community to work as closely as possible with us to determine what the best approach is here. That includes Keycloakify.

**Reply by ssilvert** [reply]
@garronej I agree with @jonkoops  on this.

I'll be happy to have a meeting some time.  We probably will convert the login theme to React at some point, but it's going to be awhile before we get to that.  So the meeting on that topic might be a little premature.  The main thing we are planning for login right now is to convert to PatternFly 5.

**Reply by jonkoops** [reply]
Note that the login is also the most used screen for any users of Keycloak, so having a good solution in terms of performance is paramount. So we might have to opt for more progressive enhancement, where part (or most) of the application is rendered on the server, and then the client simply enhances the parts that would be dynamic.

We'll have to see what would be the best solution to fit in here technically, as customization of this screen is also extremely important.

**Reply by PinPinIre** [reply]
@jonkoops This might be more of a feature request. I know that KC uses Patternfly but it would be really helpful if there was a way to switch out the components that get rendered. It's very common for companies to have a design system used across their entire product/platform. If we could use some of our design system components it would get us most of the way to customising Login.

Eg. If Login Buttons have a API, as long as our components conform to this API we can switch out the Patternfly buttons with our own.

**Reply by jonkoops** [reply]
Yes, this is indeed something we need to look at. Extending React components is not an easy task to accomplish, whereas the current setup with templates is rather easily customized. Perhaps this is something we can take as a learning from Keycloakify.

**Reply by garronej** [reply]
> So we might have to opt for more progressive enhancement, where part (or most) of the application is rendered on the server, and then the client simply enhances the parts that would be dynamic.  

While Keycloakify doesn't yet offer server-side rendering, we've optimized our loading process to download only the resources essential for rendering the specific page you're visiting.  
I mention this to emphasize that performance is a significant concern for us.

**Reply by jonkoops** [reply]
Yes, we'll have to see what is a good fit for us. Ideally we'd like to use our existing stack of React and PatternFly for the sake of consistency. However, it might be that this stack is deemed to heavy for something like the login theme, in which case we'd have to consider alternatives.

**Reply by PinPinIre** [reply]
My apologies for resurrecting quite an old thread. It's been 3 years and I am evaluating the login theme again in the latest version of Keycloak. It looks like the latest Keycloak updated the login theme but it is still a pure FTL implementation. 

The rest of our application is written in React with our own design system and component library. Ideally we would like to use the out of the box solution from Keycloak (so no 3rd party library between us and Keycloak) and make it look and feel as close as possible to our own components as possible. I was hoping to get some clarification before we invest in redoing our login from scratch. 

@ssilvert @jonkoops (sorry for the @ if you are no longer involved in the KC project) in the past 3 years has there been an update on the roadmap to migrate the KC login to use React and Patternfly like the other UI's in Keycloak?

**Reply by garronej** [reply]
Keycloakify now features a starter theme based on Shadcn/Tailwind: https://docs.keycloakify.dev/starter-themes/shadcn-ui-tailwind

**Reply by PinPinIre** [reply]
For our use case Keycloakify is not an option sadly.

**Reply by garronej** [reply]
Ok, then you might want to take a look at what @kathari00 has been working on:

https://www.linkedin.com/feed/update/urn:li:activity:7431277873300226048/?originTrackingId=xS0%2BxF226hRX62vy4V7U%2FA%3D%3D  

She’s a longtime Keycloakify contributor and has built a separate project (outside of Keycloakify) focused on creating simple themes using a WYSIWYG editor. It generates a theme bundle that’s closer to what you would expect if you had written the theme manually.

The idea is that it requires less long-term commitment to maintaining a third-party tool, while still enabling you to get something good enough, faster.

Out of curiosity, what’s the specific blocker preventing you from using Keycloakify?

**Comment by garronej** [comment]
Hi @ssilvert, hi all,  

Just a quick update on [Keycloakify](https://keycloakify.dev)  as v7 has just been released.  

Thanks to the sponsorship money of Corsair and the active community of contributors, we now have:  
- Windows support (previously it was Unix only, Window users had to use WSL)
- [Account theme support](https://youtu.be/WMyGZNHQkjU?t=202)  
- The DX is much better in many ways, there is a command to eject pages, you can pick the page you want to customize at the component level. It's now easy to see what classes are available to be overloaded.  
- [A storybook](https://storybook.keycloakify.dev/storybook/?path=/story/theme-pages-login-login--default) that enable the designers to explore the default user-facing pages. (We are in the process of making it look better.)  

The community is currently working [on bringing Vite support](https://github.com/keycloakify/keycloakify/pull/275) (currently Keycloakify only works with Webpack)

**Comment by ssilvert** [comment]
Congrats on the release!

**Comment by garronej** [comment]
Hello @ssilvert,

I wanted to provide a quick update to let you know that we've released [Keycloakify](https://github.com/keycloakify/keycloakify) version 9, which now includes complete Vite support.  
Congratulations on the version 24 release; it's great that the declarative user profile is now enabled by default!
We've observed some changes in the FreeMarker templates and plan to integrate these changes into our React components soon.

Please feel free to reach out for anything at all; we'd love to demonstrate the capabilities of Keycloakify sometime if you're interested!

Best regards,

**Reply by AndreyYolkin** [reply]
Wow, seems like new vite plugin doesn't heavely rely on react, it opens the opportunity to make another non-react vite-based starters, am I right?

**Reply by garronej** [reply]
Totally. Actually there's two main component in Keycloakify: the build tool/bundler and the React component library.   
The build tool is mostly agnostique of the UI framwork you're using.  
The component library on the other end is specific to react but is meerly a direct translation to the base FTL template.

**Reply by AndreyYolkin** [reply]
<img width="1728" alt="image" src="https://github.com/keycloak/keycloak/assets/11289484/973b5ba6-7928-4fbf-aca2-66c9c51412d7">

It'll take more time to handle `i18n` and some utils, like `usePrepareTemplate` and `getClassName`, but in general I can confirm it works with vue

**Reply by garronej** [reply]
@AndreyYolkin This is so hype!  
Don't hesitate to join [our discord](https://discord.gg/kYFZG7fQmn) if you want to push this futher!

**Reply by petitkriket** [reply]
Hello, it's a pity Vue.js is not supported. Any change @AndreyYolkin could share is PoC ? I would love to try convert it and contribute here so Vue.js can be used with Keycloakify.

**Reply by garronej** [reply]
Hello @petitkriket,  

Since we already support Svelte, I believe adding Vue support wouldn’t require too much additional effort.  
The real challenge isn’t the initial implementation, but the ongoing maintenance.  

I’m already close to the limit of what I can reasonably maintain while still making improvements elsewhere.  

That said, there’s a major move to Login v2 currently in progress. Once that’s complete, we’ll be in a better position to reevaluate adding Vue support.

**Reply by petitkriket** [reply]
Totally get it. Keeping the surface of code to "manageable" is very important. Thank you for your feedback. If at some point you're up to it, i'd love to help, even with a simple QA + Test session.
